mobile/src/main/java/org/openhab/habdroid/core/connection/ConnectionAvailabilityAwareActivity.java

```diff
@@ -0,0 +1,171 @@
+package org.openhab.habdroid.core.connection;
+
+import android.app.Fragment;
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.graphics.PorterDuff;
+import android.graphics.drawable.Drawable;
+import android.os.Bundle;
+import android.support.v4.app.TaskStackBuilder;
+import android.support.v4.content.ContextCompat;
+import android.support.v7.app.AppCompatActivity;
+import android.text.TextUtils;
+import android.util.Log;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.Button;
+import android.widget.ImageView;
+import android.widget.TextView;
+
+import org.openhab.habdroid.R;
+import org.openhab.habdroid.core.connection.exception.ConnectionException;
+import org.openhab.habdroid.core.connection.exception.NetworkNotAvailableException;
+import org.openhab.habdroid.core.connection.exception.NetworkNotSupportedException;
+import org.openhab.habdroid.core.connection.exception.NoUrlInformationException;
+import org.openhab.habdroid.core.message.MessageHandler;
+import org.openhab.habdroid.ui.OpenHABPreferencesActivity;
+
+import static org.openhab.habdroid.ui.OpenHABPreferencesActivity.NO_URL_INFO_EXCEPTION_EXTRA;
+import static org.openhab.habdroid.ui.OpenHABPreferencesActivity.NO_URL_INFO_EXCEPTION_MESSAGE;
+
+public abstract class ConnectionAvailabilityAwareActivity extends AppCompatActivity {
+    private static final String TAG = ConnectionAvailabilityAwareActivity.class.getSimpleName();
+    public static final String NO_NETWORK_TAG = "noNetwork";
+
+    private final ConnectionChangeListener mConnectionChangeListener = new ConnectionChangeListener();
+
+    public Connection getConnection(int connectionType) {
+        try {
+            return ConnectionFactory.getConnection(connectionType);
+        } catch (NetworkNotAvailableException | NetworkNotSupportedException e) {
+            showNoNetworkFragment(e.getMessage());
+        } catch (NoUrlInformationException e) {
+            Intent preferencesIntent = new Intent(this, OpenHABPreferencesActivity.class);
+            preferencesIntent.putExtra(NO_URL_INFO_EXCEPTION_EXTRA, true);
+            preferencesIntent.putExtra(NO_URL_INFO_EXCEPTION_MESSAGE, e.getMessage());
+
+            TaskStackBuilder.create(this)
+                    .addNextIntentWithParentStack(preferencesIntent)
+                    .startActivities();
+        }
+        return null;
+    }
+
+    private void showNoNetworkFragment(String message) {
+        Fragment noNetworkFrament = new NoNetworkFragment();
+        Bundle bundle = new Bundle();
+        bundle.putString(NoNetworkFragment.NO_NETWORK_MESSAGE, message);
```

Comment:
I'm not sure whether it's a good idea to show an (untranslated) exception message directly in the UI. Shouldn't we rather generate a message depending on the exception type etc.? (Same applies to the preferences activity invocation above)

Reply:
The exceptions always contain a translated message. We could move the selection of the message from the creator of the exception to this handler, though, if you want.